### DISCUSSION TITLE
Add WM_NAME

### DISCUSSION THREAD

**Discussion by lazylace37** [tuurep/windowlist#3] [Ideas]
Hi!
First of all, very cool project!

I think it would be nice to display the window name instead of the class, to differentiate similar windows.  
What do you think about it?

If you are OK with this, I could start working on it.

**Comment by tuurep** [comment]
Hi, thank you! Very cool that you're interested.

Are you familiar with the script [polywins](https://github.com/uniquepointer/polywins), where they had the options:

```bash
show="window_class"
#show="window_classname"
#show="window_title"
```

So am I right that you're after the `window title` property? If so, yes that's definitely doable and we could bring back that option from polywins.

The reason why I removed it was because there are (common) cases where the window title can be extremely long, and my bar would have space for like 2 titles at once without heavy truncation.

Example:

```
_NET_WM_NAME(UTF8_STRING) = "Oxford PV sets new solar panel efficiency world record | Oxford PV — Mozilla Firefox"
```

I had a personal feeling there's no sensible use case for this option, and was waiting for someone to tell me that they do prefer it -- thus, here you are :smile: 

Important to know: in the [wmctrl](https://github.com/Conservatory/wmctrl) source code there's already a function to get window title:

https://github.com/Conservatory/wmctrl/blob/master/main.c#L1384

Would work as is as long as `gchar` -> `char` and `g_free` -> `free`

Then we need to decide, should `window_props` struct have separate fields for `class` and `title` strings, or inside `windowlist.c` only get one of these as `name` according to the config setting?

Former means you could even use both (potentially interesting idea), latter means the config should be exposed to `windowlist.c` also, as it's now only used in `main.c`

What do you think? Also - if you were thinking about something other than `WM_NAME` (=window title), let me know ;D **Edit:** Oh yeah, it's even in the title of this discussion, haha

**Reply by lazylace37** [reply]
> The reason why I removed it was because there are (common) cases where the window title can be extremely long

This is true (especially for chromium titles), but maybe with `name_max_length` set it will look nice.

> in the [wmctrl](https://github.com/Conservatory/wmctrl) source code there's already a function to get window title

That's good: in fact, I just found out the differences between `_NET_WM_NAME` and `WM_NAME` and it's good there's a function that already handles that.

> Then we need to decide, should window_props struct have separate fields for class and title strings, or inside windowlist.c only get one of these as name according to the config setting?

I saw you are planning to add window nicknames. Have you already thought about how to implement it? If we set `name` with the title we would break the mapping between the `name` and the nickname.

**Reply by tuurep** [reply]
> but maybe with `name_max_length` set it will look nice.

Yeah, if you like this, and would use it, definitely lets add the WM_NAME option :)

> I just found out the differences between _NET_WM_NAME and WM_NAME and it's good there's a function that already handles that.

Yeah, this makes it very simple for us

> I saw you are planning to add window nicknames. Have you already thought about how to implement it? If we set name with the title we would break the mapping between the name and the nickname.

Hmm, I believe it will not be a *problem* for the case where `name=WM_NAME` is on, but rather there's the question, will it be useful to you when you choose `WM_NAME`?

My plan was just to add a list of string-to-string mappings, so this could just as well work for you but now in 2 different ways:

* For programs with a static `WM_NAME`, if you'd wish it to be something else you could have a line in `window_nicknames`
* For programs with dynamic `WM_NAME`, when it is in a certain state, I want to override the name to something else
   * Example: I have PcManFM open at `/home/tuure`, `WM_NAME` is `tuure`, but I want it to be `HOME`

Well, that does make me think of one possible problem: different programs having a case where `WM_NAME` would be the same such as those filepath examples. Maybe `window_props` should in fact have both `class` and `title`, so it'd allow potential customizing `title` on a per-`class` basis, in the future?

Also, if `window_nicknames` isn't useful for the `WM_NAME` case, we can easily have it unused either by it being empty or only going through it if `name=class` is on.

BUT lets not get *too* carried away. You're giving me a lot of ideas :smile: 

Lets just add the option at this point, I'd like it to be named `name = "class" | "title"` , in `config.toml`, but feel free to suggest otherwise.

## More offtopic + ideas stuff

I totally understand you on the problem of differentiating windows that have the same class, for me it was solved by a combination of sorting the windows by position on screen, and using `rofi show window` to see the `WM_NAMES`, and also using that to switch.

I've also had ideas of using a small id for subsequent windows like, if I have only one firefox window open it would be `firefox`, but if two, they could be either:
* `firefox(1) · firefox(2)`
* `firefox(a) · firefox(b)`
* Something similar

Feel free to ignore or talk more about that if it's interesting.

**Reply by lazylace37** [reply]
> Maybe window_props should in fact have both class and title, so it'd allow potential customizing title on a per-class basis, in the future?

This could be interesting. This way, you could add a `window_nickname` to some programs by their `class`, and the rest of them get their `WM_NAME`.

> we can easily have it unused either by it being empty or only going through it if name=class is on.

Maybe this can start like this and then see what happens :smiley:. There are many possibilities.

> I've also had ideas of using a small id for subsequent windows

This is also very interesting.

---

I run a stacking WM BTW - I guess you are running a tiling WM, if you are sorting by position. In my case, I often move windows around, so the windows list doesn't always have a meaningful order.  
That's why I proposed to add the `WM_NAME` :smile: .

Anyway, if you want I can start working on it tomorrow.  
Cheers

**Reply by tuurep** [reply]
Ok, agreed! So we can replace `name` with `class` and `title` in `window_props`

> Anyway, if you want I can start working on it tomorrow.

Thanks! Sounds good.

> I guess you are running a tiling WM, if you are sorting by position.

Actually I do use a stacking WM (Openbox). I guess I have a pretty odd way to use my desktop. I heavily use keybinds to "snap" my windows to the edges (half screen / one third of screen). Kind of manual pseudo-tiling or something..

A very common scenario is that I have 2 half-screen firefox windows, left and right. Now I see very clearly: ah, it's left firefox, and right firefox :smile: and I need to 'dig' one up because another window is on top of it.

Btw with the position sorting, the list sorts in real-time (it's best to show by a video):

https://github.com/tuurep/windowlist/assets/15036238/eeede295-301a-4925-adf6-46eabbd2f5c9

I'd argue the order will always be 'meaningful', but if you have too many windows on the bottom of the stack so that they're not visible, yeah it could get too chaotic.

**Comment by tuurep** [comment]
**Slightly off-topic:**

Also polywins had the option for `classname` -- I'd like to clarify why I dropped that and it's probably not a good option:

Example `WM_CLASS` property in my pdf reader:

```
WM_CLASS(STRING) = "org.pwmt.zathura", "Zathura"
```

The first string from that is the `classname`, and second is `class`, most of the time the `class` is short and understandable, but `classname` can be something similar.... but weird and sometimes confusing. So it's like an inferior version of `class` for the purpose of this script.
